### PR TITLE
Add review policy

### PR DESCRIPTION
**Pull request by ahrtr**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

cc @fuweid @ivanvc @jmhbnz @serathius @siyuanfoundation @spzala

### CODE REVIEW

**Comment by ahrtr** [comment]
Also see https://github.com/etcd-io/etcd/issues/19987#issuecomment-2893790250

**Review by serathius — CHANGES_REQUESTED** [review]
https://github.com/etcd-io/etcd/issues/19987#issuecomment-2893949888

**Comment by ahrtr** [comment]
> [#19987 (comment)](https://github.com/etcd-io/etcd/issues/19987#issuecomment-2893949888)

Let's discuss in this PR. Which policy you disagree?

**Comment by ahrtr** [comment]
Hi @fuweid @ivanvc @jmhbnz @serathius @siyuanfoundation let's focus on the policies proposed in this PR. It's based on current status. In future when we integrate K8s's auto merge utility, we can revisit & update this guide by then.

Please share your thoughts, thx

File: Documentation/contributor-guide/review.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+## Default rule
+
+PRs should get at least two approvals (/lgtm or GitHub review approval) before merging.
+
+Notes:
+
+- Approvals should come from a maintainer, reviewer, or submodule owner familiar with the relevant code or area.
+- If there’s disagreement, maintainers should discuss and agree before merging.
+
+## Exceptions for Less Impactful PRs
```
[the remark this answers is not in the document]
Comment (ahrtr) [reply Documentation/contributor-guide/review.md:14]: > so requiring two approvals seems reasonable to me.

Yes, it's true.



> exceptions make the process harder to follow.

That's the reason why I provide a list of examples below.

[the remark this answers is not in the document]
Comment (ahrtr) [reply Documentation/contributor-guide/review.md:14]: > exceptions make the process harder to follow.

For test, it's indeed sometime harder to follow, but let's provide some flexiblities. At least, it won't break the  production code.

[the remark this answers is not in the document]
Comment (ahrtr) [reply Documentation/contributor-guide/review.md:14]: > Sometimes, flaky test cases can make it even worse.

Good point! Removed `tests` from the list.

File: Documentation/contributor-guide/review.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Review policy
```
Comment (jmhbnz) [inline Documentation/contributor-guide/review.md:1]: Rather than a separate file could we add this policy into our existing pr management document https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/triage_prs.md.

**Review by jmhbnz — APPROVED** [review]
LGTM - Thanks @ahrtr 

- The default rule of two approvals makes sense.
- An exception of single reviewer for trivial changes like minor doc tweaks makes sense.
- Completely agree we should not merge code that has not been reviewed.

File: Documentation/contributor-guide/review.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (serathius) [reply Documentation/contributor-guide/review.md:14]: https://github.com/etcd-io/etcd/issues/19987#issuecomment-2893949888

**Comment by serathius** [comment]
This PR only needlessly split the discussion.

File: Documentation/contributor-guide/review.md (hunk 2 above)
Comment (ahrtr) [reply Documentation/contributor-guide/review.md:1]: Sounds good. Will move to there later.

**Review by spzala — COMMENTED** [review]
Thanks @ahrtr !! More clarity with the review process is better. As @jmhbnz mentioned moving it to existing triage_prs.md  or https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#get-your-pull-request-reviewed sounds good.

File: Documentation/contributor-guide/review.md (hunk 2 above)
Comment (ahrtr) [reply Documentation/contributor-guide/review.md]: Updated & thx

**Comment by ahrtr** [comment]
hi team, is there any blocker to merge this PR?

**Review by fuweid — APPROVED** [review]

**Comment by ahrtr** [comment]
@ivanvc @siyuanfoundation @serathius  @spzala

**Comment by serathius** [comment]
I am not a fan of highly restrictive code review policies. While I agree that a minimum standard for code review is necessary, I don't believe that making the rules overly explicit is beneficial. These regulations typically apply only to those with merge permissions, such as project leads. Does this imply a lack of trust in our maintainers' ability to exercise sound judgment?

I advocate for a system rooted in trust and contextual decision-making. Consider the Kubernetes repository as an example. Pull requests there require both lgtm and approval. However, any member can provide an lgtm. This means an approver could theoretically merge with anyones review. In practice, this does not happen due to the community's trust-based culture. Approvers will typically use the hold command to ensure they receive feedback from the most qualified individuals.

Reflecting on the incident that prompted this discussion https://github.com/etcd-io/etcd/issues/19987#issuecomment-2894477717, I regret my decision to implement rushed fixes to Antithesis integration additional review. When myself reviewing contributions from other contributors, I overlooked certain edge cases, which led to failures in the Antithesis tests. This lead to me needing to rushly fix them. However, I am not convinced that an additional review would have prevented this, as no other maintainers were deeply familiar with that specific area of the code. My hope is that we can cultivate a  reviewers to foster a healthier maintenance process for this subproject. I have already identified two promising candidates.

Therefore, I would like to propose an alternative. Instead of documenting an arbitrary set of rules, we should align with the Kubernetes community's practices and establish a minimum review standard by delegating merging authority to the Prow bot. This should that a PR is reviewed by other member of etcd project, and use best judgement of maintainers to ask for additional review for harder PRs.

**Comment by ahrtr** [comment]
It isn't about trust, it's about process for a healthy community/project. Also it isn't a restrictive policy, we have flexibilities. Also it isn't aiming at any individual, it's blameless as human makes mistake. 

I defer to all sig leads and maintainers/reviewers' vote to move forward. cc @ivanvc @siyuanfoundation @spzala please let me know your feedback, thx





> we should align with the Kubernetes community's practices and establish a minimum review standard by delegating merging authority to the Prow bot.

It's open to discussion as a separate topic. Before that, let's make a decision on this PR.

File: Documentation/contributor-guide/triage_prs.md [hunk 3]
```diff
@@ -30,3 +30,32 @@ Reviewers are responsive in a timely fashion, but considering everyone is busy,
...
 Make sure that appropriate reviewers are added to the PR. Also, make sure that a milestone is identified. If any of these or other important labels are missing, add them. If a correct label cannot be decided, leave a comment for the maintainers to do so as needed.
+
+## Review policy
+
+To ensure code quality and shared ownership, this review policy applies to all pull requests (PRs).
+
+### Default rule
+
+PRs should get at least two approvals (/lgtm or GitHub review approval) before merging.
+
```
Comment (jberkus) [inline Documentation/contributor-guide/triage_prs.md:41]: So if we use Prow, these are going to be in the form of `/lgtm` and `/approve` commands, rather than via Github reviews.  Kubernetes policy is one "approve" and one "lgtm", and that's worked pretty well.

File: Documentation/contributor-guide/triage_prs.md [hunk 4]
```diff
@@ -30,3 +30,32 @@ Reviewers are responsive in a timely fashion, but considering everyone is busy,
...
+* Documentation
+* Comments
+
+The rule can be relaxed:
+
+* One approval is generally enough.
+
+However:
+
+* If the author is a maintainer, they should still get approval from another maintainer, reviewer, or submodule owner, even for minor changes.
```
Comment (jberkus) [inline Documentation/contributor-guide/triage_prs.md:61]: If we adopt the Kubernetes workflow, this is mostly ensured, since even if you have self-approval in a repo, you need someone else to /lgtm.

File: Documentation/contributor-guide/triage_prs.md (hunk 3 above)
Comment (ahrtr) [reply Documentation/contributor-guide/triage_prs.md:41]: There isn't a consensus to use the K8s auto merge bot for now yet.

As mentioned in https://github.com/etcd-io/etcd/pull/19988#issuecomment-3012435862, it's open to discussion separately. We can revisit this doc if eventually it's adopted.  But before that, we need to follow a policy.

**Review by spzala — APPROVED** [review]
As I mentioned in a comment earlier in this PR, https://github.com/etcd-io/etcd/pull/19988#pullrequestreview-2858011859, more clarity is good. Thanks @ahrtr !!

**Comment by ahrtr** [comment]
Merging this PR given 5/7 approvals, thx all.
